pyccel/parser/semantic.py

```diff
@@ -948,12 +948,14 @@ def _handle_function_args(self, arguments):
         """
         args  = []
         for arg in arguments:
-            a = self._visit(arg.value)
-            if isinstance(a, FunctionDef) and not isinstance(a, PyccelFunctionDef) and not a.is_semantic:
-                self._annotate_the_called_function_def(a)
             a = self._visit(arg)
-            if isinstance(a.value, StarredArguments):
-                args.extend([FunctionCallArgument(av) for av in a.value.args_var])
+            val = a.value
+            if isinstance(val, FunctionDef) and not isinstance(val, PyccelFunctionDef) and not val.is_semantic:
+                semantic_func = self._annotate_the_called_function_def(val)
+                a = FunctionCallArgument(semantic_func, keyword = a.keyword, python_ast = a.python_ast)
+
+            if isinstance(val, StarredArguments):
+                args.extend([FunctionCallArgument(av) for av in val.args_var])
```

Comment:
Could you please explain this change?

Reply:
Before this change we were visiting the syntactic node twice. The behaviour on the second pass can be slightly different and this *may* cause problems. Further if the argument is complex (e.g. an arithmetic expression) then it could be costly to (unnecessarily) visit twice.

While I was developing the double visitation caused problems however if I revert the change I see that they have now been resolved so this change is no longer necessary. The problem that I was seeing was the following:

If we have:
```python
def f() -> 'float[:]':
    ....

def g(a : 'float[:]', b : int):
    ...

if __name__ == '__main__':
    g(f(), 4)
```
Pyccel translates this to:
```python
def f() -> 'float[:]':
    ....

def g(a : 'float[:]', b : float):
    ...

if __name__ == '__main__':
    Dummy_0000 = f()
    g(Dummy_0000, 4)
```
This is done in `_visit_FunctionCallArgument` via the visit of the value. When we visit a `FunctionCall` (`f()`), if it does not have an `Assign` in its user nodes then we create the assignment and return the `Dummy` variable. This means that `_visit_FunctionCallArgument(f())` returns `Dummy_0000`. Somehow on the second pass it was collecting the semantic `f()` which already had an `Assign` in its user nodes. As a result it was not creating the assign in the call where the result is saved. I don't remember if this meant that the dummy was not allocated or if the function call was simply not printed.

As I say I can no longer reproduce this problem so if you prefer the changes can be reverted. However I think it is better to avoid visiting an object of unknown complexity multiple times. This could be costly in the case of something like:
```python
g(1+b**2+a*(4+c*d))
```

Reply:
Thanks for the explanation. I agree, we should avoid visiting the node twice.